### DISCUSSION TITLE
How to set Ai Key for custom model

### DISCUSSION THREAD

**Discussion by wss29** [eclipse-theia/theia#14333] [Q&A] [answered]
Hi
   1. I set a custom model as follows
![image](https://github.com/user-attachments/assets/260f2b15-21e4-451a-bdba-fea51f30d7d8)
   2. My model also needs an AI key for work. I config it 
![image](https://github.com/user-attachments/assets/6d54720d-33c3-4ede-a70a-129c30d6d054)
![image](https://github.com/user-attachments/assets/9fdb33f7-a145-49d7-a887-82a2e3ea2026)
   3.but when I run @Universal hello, I get the error "Incorrect API key provided"
![image](https://github.com/user-attachments/assets/a8a90ad8-43e8-4e4f-9443-da89b3b015f3)
![image](https://github.com/user-attachments/assets/b0373ee3-c7c8-49f2-b881-214545977046)
   4. I find it cause by follow code in openai-language-model.ts, when url exist, apikey will be set to 'no-key', 
![image](https://github.com/user-attachments/assets/96e79356-b193-4d97-9d44-a5550d99a2dc)
   5 I change code as follow and tried my custom model .it works.
![image](https://github.com/user-attachments/assets/077e8a55-b481-4fe7-83c9-6ffb057aa2e2)
  6 Is there any solution can set API key for custom model? or I should develop an extension for my custom model

**Comment by sdirix** [comment, marked as answer]
Hi @wss29, we recently merged support for setting custom keys for custom models, see https://github.com/eclipse-theia/theia/pull/14299 and https://github.com/eclipse-theia/theia/issues/14288.

This enhancement will be in the next release, i.e. Theia `1.55`
